src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.KeepAlive.cs

```diff
@@ -0,0 +1,316 @@
+
+                this.Observe(
+                    SendPingAsync(pingPayload));
+            }
+        }
+
+        private async ValueTask SendPingAsync(long pingPayload)
+        {
+            Debug.Assert(_keepAlivePingState != null);
+
+            byte[] pingPayloadBuffer = ArrayPool<byte>.Shared.Rent(sizeof(long));
+            BinaryPrimitives.WriteInt64BigEndian(pingPayloadBuffer, pingPayload);
+            try
+            {
+                await TrySendKeepAliveFrameAsync(
+                    MessageOpcode.Ping,
+                    pingPayloadBuffer.AsMemory(0, sizeof(long)))
+                    .ConfigureAwait(false);
+
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.KeepAlivePingSent(this, pingPayload);
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(pingPayloadBuffer);
+            }
+        }
+
+        private void OnDataReceived(int bytesRead)
+        {
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this);
+
+            if (_keepAlivePingState != null && bytesRead > 0)
+            {
+                _keepAlivePingState.OnDataReceived();
+            }
+        }
+
+        private void ThrowIfDisposedOrKeepAliveFaulted()
+        {
+            Debug.Assert(_keepAlivePingState is not null);
+
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this, $"_disposed={_disposed}, _state={_state}, _keepAlivePingState.Exception={_keepAlivePingState.Exception?.Message}");
+
+            if (_disposed && _keepAlivePingState.Exception is not null)
+            {
+                // If Exception is not null, it triggered the abort which also disposed the websocket
+                // We only save the Exception if it actually triggered the abort
+                throw new OperationCanceledException(nameof(WebSocketState.Aborted), _keepAlivePingState.Exception);
+            }
+
+            ObjectDisposedException.ThrowIf(_disposed, this);
+        }
+
+        private void ThrowIfInvalidStateOrKeepAliveFaulted(WebSocketState[] validStates)
+        {
+            Debug.Assert(_keepAlivePingState is not null);
+
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this, $"_disposed={_disposed}, _state={_state}, _keepAlivePingState.Exception={_keepAlivePingState.Exception?.Message}");
+
+            try
+            {
+                WebSocketValidate.ThrowIfInvalidState(_state, _disposed, validStates);
+            }
+            catch (Exception exc) when (_disposed && _keepAlivePingState.Exception is not null)
+            {
+                // If Exception is not null, it triggered the abort which also disposed the websocket
+                // We only save the Exception if it actually triggered the abort
+                if (exc is ObjectDisposedException ode && ode.ObjectName == typeof(ManagedWebSocket).FullName)
```

Comment:
Good catch, `WebSocketValidate.ThrowIfInvalidState` seems to throw ODE with type of `WebSocket` only, so maybe the intent is to avoid catching that one.

However, I don't see any other possibility of different ODE being thrown so I am not sure what the intent was here.